### ISSUE TITLE
max_input_voltage leads to drop in precision

### ISSUE THREAD

**Issue by Philippe-Drolet** [coreylammie/MemTorch#80]
Hello, 

I am converting a single linear layer NN using MNIST using the following parameters: 

patch_model(copy.deepcopy(network),
                                    memristor_model=memtorch.bh.memristor.VTEAM,
                                    memristor_model_params={'time_series_resolution': 1e-10},
                                    module_parameters_to_patch=[torch.nn.Linear],
                                    mapping_routine=naive_map,
                                    transistor=True,
                                    programming_routine= memtorch.bh.crossbar.Program.naive_program,
                                    tile_shape=None,
                                    scheme=memtorch.bh.Scheme.DoubleColumn,
                                    p_l = None,
                                    max_input_voltage=0.2,
                                    ADC_resolution=None,
                                    ADC_overflow_rate=None,
                                    random_initialization=False,
                                    quant_method=None)

Which leads to a constant drop in precision when testing on a pretrained MNIST NN. Changing max_input_voltage to 0.00001V or to 10000V does not change the nominal value of this drop. The MNIST NN has been normalized before training. I looked everywhere in the code where max_input_voltage appears but I cannot find the cause for this. In the tuning, the coefficient of determination is 1.00. I tried changing the torch.manual_seed value to see if the tuning was the problem and it did make a difference but of only about 0.2% on the precision .

10:51:28.396 |INFO    | Test overall accuracy: 90.63% # First test with our own NN
10:51:35.883 |INFO    | Completed network testing memristors-based in 502ms
10:51:35.884 |INFO    | Test overall accuracy: 89.24% #Test with converted NN

Tuned bh.Linear(in_features=784, out_features=50, bias=True). Coefficient of determination: 1.000000 [56.404373, 0.000000]
Tuned bh.Linear(in_features=50, out_features=10, bias=True). Coefficient of determination: 1.000000 [157.618051, 0.000000]

Do you have any idea where the problem could be coming from? Could the way I have normalized the dataset be of cause or is it the nature of the tuning? 
Thank you,

**Comment by coreylammie** [comment]
Hi @Philippe-Drolet,

Currently, if `max_input_voltage` is not None, inputs are scaled between +/- `max_input_voltage`. Please see the logic used for `memtorch.mn.Linear` below (this is the same for all layer types) that is causing the drop in accuracy:

https://github.com/coreylammie/MemTorch/blob/8b742c60cf69ebd6e68cd0c4c55118998abf4226/memtorch/mn/Linear.py#L156-L170

I agree that this behaviour is sub-optimal, and could definitely be improved. Namely:
1. This logic should be modularized for all layer types. 
3. Functionality should be added to define arbitrary scaling routines/logic (defined as external and/or lambda functions).
2. For the default routine, if `torch.amax(torch.abs(input))` is less than or equal to `max_input_voltage`, scaling does not need to be performed, unless explicitly specified using a `scaling` boolean input argument. 

I will add this functionality now and update this post with some documentation once complete.

Kind Regards,

Corey.

**Comment by coreylammie** [comment]
The aforementioned functionality has successfully been added and merged to `master` in #81.

Customizable methods described externally or in `memtorch.map.Input` now scale inputs prior to presentation. Using `memtorch.map.Input.naive_scale`, force scaling (for inputs smaller than or equal to `max_input_voltage`) can be enabled/disabled as follows:

```
import torch
import memtorch
from memtorch.map.Input import naive_scale
from memtorch.mn.Module import patch_model
import copy


class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.linear = torch.nn.Linear(100, 100)

    def forward(self, x):
        return torch.functional.relu(self.linear(x))


memristor_model = memtorch.bh.memristor.VTEAM
memristor_model_params = {}
module_parameters_to_patch = [torch.nn.Linear]
tile_shape = (8, 8)
max_input_voltage = 2.0
scaling_routine = naive_scale
ADC_resolution = 8
ADC_overflow_rate = 0.0
quant_method = "linear"

device = torch.device("cpu" if "cpu" in memtorch.__version__ else "cuda")
# force_scale -> False
patched_network_force_scale_disabled = patch_model(
    Net().to(device),
    memristor_model=memristor_model,
    memristor_model_params=memristor_model_params,
    module_parameters_to_patch=module_parameters_to_patch,
    tile_shape=tile_shape,
    max_input_voltage=max_input_voltage,
    scaling_routine=scaling_routine,
    scaling_routine_params={"force_scale": False},
    ADC_resolution=ADC_resolution,
    ADC_overflow_rate=ADC_overflow_rate,
    quant_method=quant_method,
)
patched_network_force_scale_disabled.tune_()
# force_scale -> True
patched_network_force_scale_enabled = patch_model(
    Net().to(device),
    memristor_model=memristor_model,
    memristor_model_params=memristor_model_params,
    module_parameters_to_patch=module_parameters_to_patch,
    tile_shape=tile_shape,
    max_input_voltage=max_input_voltage,
    scaling_routine=scaling_routine,
    scaling_routine_params={"force_scale": True},
    ADC_resolution=ADC_resolution,
    ADC_overflow_rate=ADC_overflow_rate,
    quant_method=quant_method,
)
patched_network_force_scale_enabled.tune_()
```

Hopefully this helps!
